yt/yt/ytlib/table_client/chunk_slice.cpp

```diff
@@ -406,15 +406,59 @@ void ToProto(
 
 ////////////////////////////////////////////////////////////////////////////////
 
+namespace {
+
+////////////////////////////////////////////////////////////////////////////////
+
+std::optional<THashSet<int>> GetBlockFilter(
+    const NChunkClient::NProto::TChunkMeta& chunkMeta,
+    const std::optional<std::vector<TColumnStableName>>& columnStableNames)
+{
+    if (columnStableNames) {
+        TNameTablePtr nameTable;
+        if (auto nameTableExt = FindProtoExtension<TNameTableExt>(chunkMeta.extensions())) {
+            nameTable = FromProto<TNameTablePtr>(*nameTableExt);
+        } else if (auto schemaExt = FindProtoExtension<TTableSchemaExt>(chunkMeta.extensions())) {
+            nameTable = TNameTable::FromSchemaStable(FromProto<TTableSchema>(*schemaExt));
+        } else {
+            return {};
+        }
+
+        THashSet<int> blockFilter;
+
+        auto columnMetaExt = FindProtoExtension<TColumnMetaExt>(chunkMeta.extensions());
```

Comment:
What happens if chunk is not columnar here?

Reply:
Nothing good :)
I think I intended to have an if after `FindProtoExtension` here. Added it.

Meanwhile, I realized another caveat: the way we are accounting for columns is not exactly the same as with columnar statistics. If I have a block that contains the columns `a`, `b` and `c` in any format and I am only reading column `a`, the columnar statistics factor will give me an approximation on the weight of the single column `a`, whereas the current logic will give me the size of the whole block. The worst case scenario here happens for regular non-columnar tables, where by using the new option option we will lose the benefits we had with columnar statistics.

The annoying part is that for columnar tables we either have to rely on columnar statistics completely (which might be quite inexact), without matching block by columns, or perform additional logic which would approximate only the relevant parts of each block by also looking into columnar statistics. 
For non-columnar tables we could fallback and **use** columnar statistics alongside chunk slice size statistics, but it seems very inconsistent with explicitly **not** using it for columnar tables. 

Thus, I decided to solidify the fact that block selectivity factors override column selectivity factors. See the new comments and logic in `GetSelectivityFactor` and `ApplySelectivityFactors`.
In the future we can improve `TChunkSliceSizeFetcher` to handle column selection better. Maybe it can be generalized across different fetchers and used inside the regular columnar statistics fetcher and inside  `TChunkSliceFetcher` as well.

But it should be noted that in some cases `use_chunk_slice_statistics = true` for non-columnar tables might produce different ("worse") job partitioning than `use_columnar_statistics = true`.

The curious part of all of this is that the jobs do actually read blocks as a whole :)